File: backend/modules/evaluation/domain/entity/evaluator.go
// Copyright (c) 2025 coze-dev Authors
// SPDX-License-Identifier: Apache-2.0

package entity

type Evaluator struct {
	ID             int64
	SpaceID        int64
	Name           string
	Description    string
	DraftSubmitted bool
	EvaluatorType  EvaluatorType
	LatestVersion  string
	BaseInfo       *BaseInfo

	Builtin               bool
	EvaluatorInfo         *EvaluatorInfo
	BuiltinVisibleVersion string
	BoxType               EvaluatorBoxType                                      `json:"box_type"`
	Tags                  map[EvaluatorTagLangType]map[EvaluatorTagKey][]string `json:"tags"`

	PromptEvaluatorVersion    *PromptEvaluatorVersion
	CodeEvaluatorVersion      *CodeEvaluatorVersion
	CustomRPCEvaluatorVersion *CustomRPCEvaluatorVersion
	AgentEvaluatorVersion     *AgentEvaluatorVersion
}

type EvaluatorInfo struct {
	Benchmark     *string `json:"benchmark"`
	Vendor        *string `json:"vendor"`
	VendorURL     *string `json:"vendor_url"`
	UserManualURL *string `json:"user_manual_url"`
}

type EvaluatorBoxType int64

const (
	EvaluatorBoxTypeWhite EvaluatorBoxType = 1
	EvaluatorBoxTypeBlack EvaluatorBoxType = 2
)

type EvaluatorType int64

const (
	EvaluatorTypePrompt    EvaluatorType = 1
	EvaluatorTypeCode      EvaluatorType = 2
	EvaluatorTypeCustomRPC EvaluatorType = 3
	EvaluatorTypeAgent     EvaluatorType = 4
)

var EvaluatorTypeSet = map[EvaluatorType]struct{}{
	EvaluatorTypePrompt:    {},
	EvaluatorTypeCode:      {},
	EvaluatorTypeCustomRPC: {},
	EvaluatorTypeAgent:     {},
}

func (e *Evaluator) IsAsync() bool {
	return e.EvaluatorType == EvaluatorTypeAgent
}

// UpdateEvaluatorMetaRequest 用于更新评估器元信息的参数
type UpdateEvaluatorMetaRequest struct {
	ID                    int64
	SpaceID               int64
	Name                  *string
	Description           *string
	Builtin               *bool
	EvaluatorInfo         *EvaluatorInfo
	BuiltinVisibleVersion *string
	BoxType               *EvaluatorBoxType
	UpdatedBy             string
}

// GetEvaluatorVersionID 获取评估器版本ID
func (e *Evaluator) GetEvaluatorVersionID() int64 {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			return e.PromptEvaluatorVersion.GetID()
		}
	case EvaluatorTypeCode:
		if e.CodeEvaluatorVersion != nil {
			return e.CodeEvaluatorVersion.GetID()
		}
	case EvaluatorTypeCustomRPC:
		if e.CustomRPCEvaluatorVersion != nil {
			return e.CustomRPCEvaluatorVersion.GetID()
		}
	case EvaluatorTypeAgent:
		if e.AgentEvaluatorVersion != nil {
			return e.AgentEvaluatorVersion.GetID()
		}
	default:
		return 0
	}
	return 0
}

// GetVersion 获取评估器版本号
func (e *Evaluator) GetVersion() string {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			return e.PromptEvaluatorVersion.GetVersion()
		}
	case EvaluatorTypeCode:
		if e.CodeEvaluatorVersion != nil {
			return e.CodeEvaluatorVersion.GetVersion()
		}
	case EvaluatorTypeCustomRPC:
		if e.CustomRPCEvaluatorVersion != nil {
			return e.CustomRPCEvaluatorVersion.GetVersion()
		}
	case EvaluatorTypeAgent:
		if e.AgentEvaluatorVersion != nil {
			return e.AgentEvaluatorVersion.GetVersion()
		}
	default:
		return ""
	}
	return ""
}

// GetEvaluatorID 获取评估器ID
func (e *Evaluator) GetEvaluatorID() int64 {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			return e.PromptEvaluatorVersion.GetEvaluatorID()
		}
	case EvaluatorTypeCode:
		if e.CodeEvaluatorVersion != nil {
			return e.CodeEvaluatorVersion.GetEvaluatorID()
		}
	case EvaluatorTypeCustomRPC:
		if e.CustomRPCEvaluatorVersion != nil {
			return e.CustomRPCEvaluatorVersion.GetEvaluatorID()
		}
	case EvaluatorTypeAgent:
		if e.AgentEvaluatorVersion != nil {
			return e.AgentEvaluatorVersion.GetEvaluatorID()
		}
	default:
		return 0
	}
	return 0
}

// GetSpaceID 获取空间ID
func (e *Evaluator) GetSpaceID() int64 {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			return e.PromptEvaluatorVersion.GetSpaceID()
		}
	case EvaluatorTypeCode:
		if e.CodeEvaluatorVersion != nil {
			return e.CodeEvaluatorVersion.GetSpaceID()
		}
	case EvaluatorTypeCustomRPC:
		if e.CustomRPCEvaluatorVersion != nil {
			return e.CustomRPCEvaluatorVersion.GetSpaceID()
		}
	case EvaluatorTypeAgent:
		if e.AgentEvaluatorVersion != nil {
			return e.AgentEvaluatorVersion.GetSpaceID()
		}
	default:
		return 0
	}
	return 0
}

// GetEvaluatorDescription 获取评估器描述
func (e *Evaluator) GetEvaluatorDescription() string {
	return e.Description
}

// GetEvaluatorVersionDescription 获取评估器版本描述
func (e *Evaluator) GetEvaluatorVersionDescription() string {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			return e.PromptEvaluatorVersion.GetDescription()
		}
	case EvaluatorTypeCode:
		if e.CodeEvaluatorVersion != nil {
			return e.CodeEvaluatorVersion.GetDescription()
		}
	case EvaluatorTypeCustomRPC:
		if e.CustomRPCEvaluatorVersion != nil {
			return e.CustomRPCEvaluatorVersion.GetDescription()
		}
	case EvaluatorTypeAgent:
		if e.AgentEvaluatorVersion != nil {
			return e.AgentEvaluatorVersion.GetDescription()
		}
	default:
		return ""
	}
	return ""
}

// GetBaseInfo 获取基础信息
func (e *Evaluator) GetBaseInfo() *BaseInfo {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			return e.PromptEvaluatorVersion.GetBaseInfo()
		}
	case EvaluatorTypeCode:
		if e.CodeEvaluatorVersion != nil {
			return e.CodeEvaluatorVersion.GetBaseInfo()
		}
	case EvaluatorTypeCustomRPC:
		if e.CustomRPCEvaluatorVersion != nil {
			return e.CustomRPCEvaluatorVersion.GetBaseInfo()
		}
	case EvaluatorTypeAgent:
		if e.AgentEvaluatorVersion != nil {
			return e.AgentEvaluatorVersion.GetBaseInfo()
		}
	default:
		return nil
	}
	return nil
}

// GetPromptTemplateKey 获取提示模板键
func (e *Evaluator) GetPromptTemplateKey() string {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			return e.PromptEvaluatorVersion.GetPromptTemplateKey()
		}
	default:
		return ""
	}
	return ""
}

// GetModelConfig 获取模型配置
func (e *Evaluator) GetModelConfig() *ModelConfig {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			return e.PromptEvaluatorVersion.GetModelConfig()
		}
	default:
		return nil
	}
	return nil
}

// ValidateInput 验证输入数据
func (e *Evaluator) ValidateInput(input *EvaluatorInputData) error {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			return e.PromptEvaluatorVersion.ValidateInput(input)
		}
	case EvaluatorTypeCode:
		if e.CodeEvaluatorVersion != nil {
			return e.CodeEvaluatorVersion.ValidateInput(input)
		}
	case EvaluatorTypeCustomRPC:
		if e.CustomRPCEvaluatorVersion != nil {
			return e.CustomRPCEvaluatorVersion.ValidateInput(input)
		}
	case EvaluatorTypeAgent:
		if e.AgentEvaluatorVersion != nil {
			return e.AgentEvaluatorVersion.ValidateInput(input)
		}
	default:
		return nil
	}
	return nil
}

// ValidateBaseInfo 校验评估器基本信息
func (e *Evaluator) ValidateBaseInfo() error {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			return e.PromptEvaluatorVersion.ValidateBaseInfo()
		}
	case EvaluatorTypeCode:
		if e.CodeEvaluatorVersion != nil {
			return e.CodeEvaluatorVersion.ValidateBaseInfo()
		}
	case EvaluatorTypeCustomRPC:
		if e.CustomRPCEvaluatorVersion != nil {
			return e.CustomRPCEvaluatorVersion.ValidateBaseInfo()
		}
	case EvaluatorTypeAgent:
		if e.AgentEvaluatorVersion != nil {
			return e.AgentEvaluatorVersion.ValidateBaseInfo()
		}
	default:
		return nil
	}
	return nil
}

// SetEvaluatorVersionID 设置评估器版本ID
func (e *Evaluator) SetEvaluatorVersionID(id int64) {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			e.PromptEvaluatorVersion.SetID(id)
		}
	case EvaluatorTypeCode:
		if e.CodeEvaluatorVersion != nil {
			e.CodeEvaluatorVersion.SetID(id)
		}
	case EvaluatorTypeCustomRPC:
		if e.CustomRPCEvaluatorVersion != nil {
			e.CustomRPCEvaluatorVersion.SetID(id)
		}
	case EvaluatorTypeAgent:
		if e.AgentEvaluatorVersion != nil {
			e.AgentEvaluatorVersion.SetID(id)
		}
	default:
		return
	}
}

// SetVersion 设置版本号
func (e *Evaluator) SetVersion(version string) {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			e.PromptEvaluatorVersion.SetVersion(version)
		}
	case EvaluatorTypeCode:
		if e.CodeEvaluatorVersion != nil {
			e.CodeEvaluatorVersion.SetVersion(version)
		}
	case EvaluatorTypeCustomRPC:
		if e.CustomRPCEvaluatorVersion != nil {
			e.CustomRPCEvaluatorVersion.SetVersion(version)
		}
	case EvaluatorTypeAgent:
		if e.AgentEvaluatorVersion != nil {
			e.AgentEvaluatorVersion.SetVersion(version)
		}
	default:
		return
	}
}

// SetEvaluatorDescription 设置评估器描述
func (e *Evaluator) SetEvaluatorDescription(description string) {
	e.Description = description
}

// SetEvaluatorVersionDescription 设置评估器版本描述
func (e *Evaluator) SetEvaluatorVersionDescription(description string) {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			e.PromptEvaluatorVersion.SetDescription(description)
		}
	case EvaluatorTypeCode:
		if e.CodeEvaluatorVersion != nil {
			e.CodeEvaluatorVersion.SetDescription(description)
		}
	case EvaluatorTypeCustomRPC:
		if e.CustomRPCEvaluatorVersion != nil {
			e.CustomRPCEvaluatorVersion.SetDescription(description)
		}
	case EvaluatorTypeAgent:
		if e.AgentEvaluatorVersion != nil {
			e.AgentEvaluatorVersion.SetDescription(description)
		}
	default:
		return
	}
}

// SetBaseInfo 设置基础信息
func (e *Evaluator) SetBaseInfo(baseInfo *BaseInfo) {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			e.PromptEvaluatorVersion.SetBaseInfo(baseInfo)
		}
	case EvaluatorTypeCode:
		if e.CodeEvaluatorVersion != nil {
			e.CodeEvaluatorVersion.SetBaseInfo(baseInfo)
		}
	case EvaluatorTypeCustomRPC:
		if e.CustomRPCEvaluatorVersion != nil {
			e.CustomRPCEvaluatorVersion.SetBaseInfo(baseInfo)
		}
	case EvaluatorTypeAgent:
		if e.AgentEvaluatorVersion != nil {
			e.AgentEvaluatorVersion.SetBaseInfo(baseInfo)
		}
	default:
		return
	}
}

// SetTools 设置工具
func (e *Evaluator) SetTools(tools []*Tool) {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			e.PromptEvaluatorVersion.SetTools(tools)
		}
	default:
		return
	}
}

// SetPromptSuffix 设置提示后缀
func (e *Evaluator) SetPromptSuffix(promptSuffix string) {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			e.PromptEvaluatorVersion.SetPromptSuffix(promptSuffix)
		}
	default:
		return
	}
}

// SetParseType 设置解析类型
func (e *Evaluator) SetParseType(parseType ParseType) {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			e.PromptEvaluatorVersion.SetParseType(parseType)
		}
	default:
		return
	}
}

// SetEvaluatorID 设置评估器ID
func (e *Evaluator) SetEvaluatorID(evaluatorID int64) {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			e.PromptEvaluatorVersion.SetEvaluatorID(evaluatorID)
		}
	case EvaluatorTypeCode:
		if e.CodeEvaluatorVersion != nil {
			e.CodeEvaluatorVersion.SetEvaluatorID(evaluatorID)
		}
	case EvaluatorTypeCustomRPC:
		if e.CustomRPCEvaluatorVersion != nil {
			e.CustomRPCEvaluatorVersion.SetEvaluatorID(evaluatorID)
		}
	case EvaluatorTypeAgent:
		if e.AgentEvaluatorVersion != nil {
			e.AgentEvaluatorVersion.SetEvaluatorID(evaluatorID)
		}
	default:
		return
	}
}

// SetSpaceID 设置空间ID
func (e *Evaluator) SetSpaceID(spaceID int64) {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			e.PromptEvaluatorVersion.SetSpaceID(spaceID)
		}
	case EvaluatorTypeCode:
		if e.CodeEvaluatorVersion != nil {
			e.CodeEvaluatorVersion.SetSpaceID(spaceID)
		}
	case EvaluatorTypeCustomRPC:
		if e.CustomRPCEvaluatorVersion != nil {
			e.CustomRPCEvaluatorVersion.SetSpaceID(spaceID)
		}
	case EvaluatorTypeAgent:
		if e.AgentEvaluatorVersion != nil {
			e.AgentEvaluatorVersion.SetSpaceID(spaceID)
		}
	default:
		return
	}
}

func (e *Evaluator) SetEvaluatorVersion(version *Evaluator) {
	if version == nil {
		return
	}
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		e.PromptEvaluatorVersion = version.PromptEvaluatorVersion
	case EvaluatorTypeCode:
		e.CodeEvaluatorVersion = version.CodeEvaluatorVersion
	case EvaluatorTypeCustomRPC:
		e.CustomRPCEvaluatorVersion = version.CustomRPCEvaluatorVersion
	case EvaluatorTypeAgent:
		e.AgentEvaluatorVersion = version.AgentEvaluatorVersion
	default:
		return
	}
}

func (e *Evaluator) GetRateLimit() *RateLimit {
	switch e.EvaluatorType {
	case EvaluatorTypeCustomRPC:
		if e.CustomRPCEvaluatorVersion != nil {
			return e.CustomRPCEvaluatorVersion.RateLimit
		}
	}
	return nil
}

func (e *Evaluator) GetInputSchemas() []*ArgsSchema {
	switch e.EvaluatorType {
	case EvaluatorTypePrompt:
		if e.PromptEvaluatorVersion != nil {
			return e.PromptEvaluatorVersion.InputSchemas
		}
	case EvaluatorTypeCode:
		return nil
	case EvaluatorTypeCustomRPC:
		if e.CustomRPCEvaluatorVersion != nil {
			return e.CustomRPCEvaluatorVersion.InputSchemas
		}
	case EvaluatorTypeAgent:
		if e.AgentEvaluatorVersion != nil {
			return e.AgentEvaluatorVersion.InputSchemas
		}
	}
	return nil
}

type EvaluatorRunConfig struct {
	Env                   *string       `json:"env,omitempty"`
	EvaluatorRuntimeParam *RuntimeParam `json:"evaluator_runtime_param,omitempty"`
}

File: frontend/packages/loop-base/api-schema/src/api/idl/evaluation/domain/evaluator.ts
// Copyright (c) 2025 coze-dev Authors
// SPDX-License-Identifier: Apache-2.0
import * as runtime from './../../llm/domain/runtime';
export { runtime };
import * as common from './common';
export { common };
export enum EvaluatorType {
  Prompt = 1,
  Code = 2,
  CustomRPC = 3,
}
export enum LanguageType {
  Python = "Python",
  /** 空间 */
  JS = "JS",
}
export enum PromptSourceType {
  BuiltinTemplate = 1,
  LoopPrompt = 2,
  Custom = 3,
}
export enum ToolType {
  Function = 1,
  /** for gemini native tool */
  GoogleSearch = 2,
}
export enum TemplateType {
  Prompt = 1,
  Code = 2,
}
export enum EvaluatorRunStatus {
  /** 运行状态, 异步下状态流转, 同步下只有 Success / Fail */
  Unknown = 0,
  Success = 1,
  Fail = 2,
}
export enum EvaluatorTagType {
  Evaluator = "Evaluator",
  Template = "Template",
}
export enum EvaluatorTagLangType {
  Zh = "zh-CN",
  En = "en-US",
}
/** Evaluator筛选字段 */
export enum EvaluatorTagKey {
  Category = "Category",
  /** 类型筛选 (LLM/Code) */
  TargetType = "TargetType",
  /** 评估对象 (文本/图片/视频等) */
  Objective = "Objective",
  /** 评估目标 (任务完成/内容质量等) */
  BusinessScenario = "BusinessScenario",
  /** 业务场景 (安全风控/AI Coding等) */
  Name = "Name",
}
/** 评估器名称 */
export enum EvaluatorBoxType {
  White = "White",
  /** 白盒 */
  Black = "Black",
}
/** 黑盒 */
export enum EvaluatorAccessProtocol {
  RPC = "rpc",
  AccessProtocol_RPCOld = "rpc_old",
  AccessProtocol_FaasHTTP = "faas_http",
  AccessProtocol_FaasHTTPOld = "faas_http_old",
}
export enum EvaluatorVersionType {
  Latest = "Latest",
  /** 最新版本 */
  BuiltinVisible = "BuiltinVisible",
}
/** 内置可见版本 */
export interface Tool {
  type: ToolType,
  function?: Function,
}
export interface Function {
  name: string,
  description?: string,
  parameters?: string,
}
export interface PromptEvaluator {
  message_list: common.Message[],
  model_config?: common.ModelConfig,
  prompt_source_type?: PromptSourceType,
  /** 最新版本中存evaluator_template_id */
  prompt_template_key?: string,
  prompt_template_name?: string,
  tools?: Tool[],
}
export interface CodeEvaluator {
  language_type?: LanguageType,
  code_content?: string,
  /** code类型评估器模板中code_template_key + language_type是唯一键；最新版本中存evaluator_template_id */
  code_template_key?: string,
  code_template_name?: string,
  lang_2_code_content?: {
    [key: string | number]: string
  },
}
export interface CustomRPCEvaluator {
  /** 自定义评估器编码，例如：EvalBot的给“代码生成-代码正确”赋予CN:480的评估器ID */
  provider_evaluator_code?: string,
  /** 本期是RPC，后续还可拓展HTTP */
  access_protocol: EvaluatorAccessProtocol,
  service_name?: string,
  cluster?: string,
  /** 执行http信息 */
  invoke_http_info?: EvaluatorHTTPInfo,
  /** ms */
  timeout?: number,
  /** 自定义评估器的限流配置 */
  rate_limit?: common.RateLimit,
  /** extra fields */
  ext?: {
    [key: string | number]: string
  },
}
export interface EvaluatorVersion {
  /** 版本id */
  id?: string,
  version?: string,
  description?: string,
  base_info?: common.BaseInfo,
  evaluator_content?: EvaluatorContent,
}
export interface EvaluatorContent {
  receive_chat_history?: boolean,
  input_schemas?: common.ArgsSchema[],
  output_schemas?: common.ArgsSchema[],
  /** 101-200 Evaluator类型 */
  prompt_evaluator?: PromptEvaluator,
  code_evaluator?: CodeEvaluator,
  custom_rpc_evaluator?: CustomRPCEvaluator,
}
/** 明确有顺序的 evaluator 与版本映射元素 */
export interface EvaluatorIDVersionItem {
  evaluator_id?: string,
  version?: string,
  run_config?: EvaluatorRunConfig,
}
export interface EvaluatorInfo {
  benchmark?: string,
  vendor?: string,
  vendor_url?: string,
  user_manual_url?: string,
}
export interface Evaluator {
  evaluator_id?: string,
  workspace_id?: string,
  evaluator_type?: EvaluatorType,
  name?: string,
  description?: string,
  draft_submitted?: boolean,
  base_info?: common.BaseInfo,
  current_version?: EvaluatorVersion,
  latest_version?: string,
  builtin?: boolean,
  evaluator_info?: EvaluatorInfo,
  builtin_visible_version?: string,
  /** 默认白盒 */
  box_type?: EvaluatorBoxType,
  tags?: {
    [key: string | number]: {
      [key: string | number]: string[]
    }
  },
}
export interface EvaluatorTemplate {
  id?: string,
  workspace_id?: string,
  evaluator_type?: EvaluatorType,
  name?: string,
  description?: string,
  /** 热度 */
  popularity?: number,
  evaluator_info?: EvaluatorInfo,
  tags?: {
    [key: string | number]: {
      [key: string | number]: string[]
    }
  },
  evaluator_content?: EvaluatorContent,
  base_info?: common.BaseInfo,
}
/** Evaluator筛选器选项 */
export interface EvaluatorFilterOption {
  /** 模糊搜索关键词，在所有tag中搜索 */
  search_keyword?: string,
  /** 筛选条件 */
  filters?: EvaluatorFilters,
}
/** Evaluator筛选条件 */
export interface EvaluatorFilters {
  /** 筛选条件列表 */
  filter_conditions?: EvaluatorFilterCondition[],
  /** 逻辑操作符 */
  logic_op?: EvaluatorFilterLogicOp,
  sub_filters?: EvaluatorFilters[],
}
/** 筛选逻辑操作符 */
export enum EvaluatorFilterLogicOp {
  Unknown = "Unknown",
  And = "And",
  /** 与操作 */
  Or = "Or",
}
/**
 * 或操作
 * Evaluator筛选条件
*/
export interface EvaluatorFilterCondition {
  /** 筛选字段 */
  tag_key: EvaluatorTagKey,
  /** 操作符 */
  operator: EvaluatorFilterOperatorType,
  /** 操作值 */
  value: string,
}
/** Evaluator筛选操作符 */
export enum EvaluatorFilterOperatorType {
  Unknown = "Unknown",
  Equal = "Equal",
  /** 等于 */
  NotEqual = "NotEqual",
  /** 不等于 */
  In = "In",
  /** 包含于 */
  NotIn = "NotIn",
  /** 不包含于 */
  Like = "Like",
  /** 模糊匹配 */
  IsNull = "IsNull",
  /** 为空 */
  IsNotNull = "IsNotNull",
}
/** 非空 */
export interface Correction {
  score?: number,
  explain?: string,
  updated_by?: string,
}
export interface EvaluatorRecord {
  id?: string,
  experiment_id?: string,
  experiment_run_id?: string,
  item_id?: string,
  turn_id?: string,
  evaluator_version_id?: string,
  trace_id?: string,
  log_id?: string,
  evaluator_input_data?: EvaluatorInputData,
  evaluator_output_data?: EvaluatorOutputData,
  status?: EvaluatorRunStatus,
  base_info?: common.BaseInfo,
  ext?: {
    [key: string | number]: string
  },
}
export interface EvaluatorOutputData {
  evaluator_result?: EvaluatorResult,
  evaluator_usage?: EvaluatorUsage,
  evaluator_run_error?: EvaluatorRunError,
  time_consuming_ms?: string,
  stdout?: string,
}
export interface EvaluatorResult {
  score?: number,
  correction?: Correction,
  reasoning?: string,
}
export interface EvaluatorUsage {
  input_tokens?: string,
  output_tokens?: string,
}
export interface EvaluatorRunError {
  code?: number,
  message?: string,
}
export interface EvaluatorInputData {
  history_messages?: common.Message[],
  input_fields?: {
    [key: string | number]: common.Content
  },
  evaluate_dataset_fields?: {
    [key: string | number]: common.Content
  },
  evaluate_target_output_fields?: {
    [key: string | number]: common.Content
  },
  ext?: {
    [key: string | number]: string
  },
}
export interface EvaluatorHTTPInfo {
  method?: EvaluatorHTTPMethod,
  path?: string,
}
export enum EvaluatorHTTPMethod {
  HTTPMethod_Get = "get",
  HTTPMethod_Post = "post",
}
export interface EvaluatorRunConfig {
  env?: string,
  evaluator_runtime_param?: common.RuntimeParam,
}